### PR TITLE
fix!: ensure deleting message does not accidentally delete unrelated notifications

### LINKED ISSUE #17052: All replies and mentions are removed in AC for receiver after sender removes the last message in community chat
### Steps to reproduce:
1. `User_A` and `User_B` are both in the same community.
2. `User_A` sends several replies or mentions to `User_B`.
3. `User_A` sends one more message and then deletes it using the 'delete for everyone' option.
4. `User_B` checks the AC.

### Actual result:
All mentions and replies in the AC are removed for `User_B`.

### Expected result:
Mentions and replies should not be removed from the AC after the last message is deleted.

### Additional info:
This issue is similar in group chats, with the exception that only replies are removed from the AC. Mentions are not removed and

### PR DESCRIPTION
**Pull request by seanstrom**
fixes https://github.com/status-im/status-mobile/issues/17052
status-mobile PR: https://github.com/status-im/status-mobile/pull/21157
this PR is a follow up to: https://github.com/status-im/status-go/pull/5520/

## The Issue

* This PR attempts to resolve an issue with notifications being deleted for users after another user deletes a message inside a community.
  * The cause of the issue seems to be related to how notifications can reference a message with the `LastMessage` field. And if a message is deleted from a community and the notifications reference that message via `LastMessage`, then the notifications are also removed/deleted
    * We can see where the notifications are removed via `LastMessage` here: https://github.com/status-im/status-go/blob/bd0464d192971a284f29cf64c97a1914124307fe/protocol/activity_center_persistence.go#L51-L70
 
### Some Context
* It seems that `LastMessage` has been used for some notifications to represent the message for a `NewOneToOne` or `NewPrivateGroupChat` notification. The other activity-center notification types do not seem to use `LastMessage` field directly
* It's also important to note that the `LastMessage` field for notifications is not stored with the notification data. It is actually derived/joined from the `chat` data via the `chatID`. 
* Based on the git commit history, it seem the `LatestMessage` checks for deleting notifications were originally meant for detecting the `NewOneToOne` and `NewPrivateGroupChat` notifications. Here's the link to the original PR: https://github.com/status-im/status-go/pull/2854
* For context here's the main usage of the notification LastMessage field in status-mobile: https://github.com/status-im/status-mobile/blob/seanstrom/skip-audio-player-tests/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L68
* And here's the the main usage of the notification LastMessage field in status-go: https://github.com/status-im/status-desktop/blob/7c184f35002886947493070400500e7e05de74b6/src/app_service/service/activity_center/dto/notification.nim#L134-L138

### This PR Solution
* This PR tries to transition the activity notifications for `NewOneToOne` and `NewPrivateGroupChats` to use the `Message` field instead of the `LastMessage` field, which should remove the need to delete notifications via the `LatestMessage` field.
* This PR also attempts to focus on detecting `NewOneToOne` or `NewPrivateGroupChat` notification types that have references to a deleted message via `LastMessage`.
  * It might not be necessary to still detect these edge cases.

### CODE REVIEW

File: protocol/activity_center_persistence.go [hunk 1]
```diff
@@ -61,7 +61,10 @@ func (db sqlitePersistence) DeleteActivityCenterNotificationForMessage(chatID st
 
 	for _, notification := range notifications {
 		if notification.LastMessage != nil && notification.LastMessage.ID == messageID {
-			withNotification(notification)
+			switch notification.Type {
+			case ActivityCenterNotificationTypeNewOneToOne, ActivityCenterNotificationTypeNewPrivateGroupChat:
+				withNotification(notification)
+			}
```
Comment (seanstrom) [inline protocol/activity_center_persistence.go]: Here we're focusing on only removing notifications that might have a custom `LastMessage` from creating a `NewOneToOne` or `NewPrivateGroupChat` notification. Although it's likely that we won't need this anymore since we no longer use the `LastMessage` field when creating those notifications.

If it makes sense to delete this `if-block` then that's what I'll do next.

File: protocol/messenger_handler.go [hunk 2]
```diff
@@ -292,7 +292,7 @@ func (m *Messenger) createMessageNotification(chat *Chat, messageState *Received
 	notification := &ActivityCenterNotification{
 		ID:          types.FromHex(chat.ID),
 		Name:        chat.Name,
-		LastMessage: message,
+		Message:     message,
```
Comment (seanstrom) [inline protocol/messenger_handler.go:298]: Here we're changing the creation of the `NewOneToOne` and `NewPrivateGroupChat` notifications to use the `Message` field instead of the `LastMessage` field.

File: protocol/activity_center_persistence_test.go [hunk 3]
```diff
@@ -228,6 +233,103 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
 	s.Require().NoError(err)
 }
 
+func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificationsForMessage_LastMessage() {
+	// docs: Create the temporary test-database that will be used to store chats, messages, and notifications
+	db, err := openTestDB()
+	s.Require().NoError(err)
```
Comment (seanstrom) [inline protocol/activity_center_persistence_test.go]: Here we're creating another test case for specifically deleting NewOneToOne and NewPrivateGroupChat notifications via matching the `LastMessage` field with the deleted message ID. However, if we decide that we no longer need to support checking the LastMessage field when deleting notifications, then we can remove this test case.

File: protocol/activity_center_persistence_test.go [hunk 4]
```diff
@@ -205,17 +205,22 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
...
+	s.Require().True(notif.Dismissed, notif.ID)
+	s.Require().True(notif.Read, notif.ID)
+
+	// Check that notifications with LastMessage.ID == messages[1].ID will remain.
+	// Unless the notification is of type NewOneToOne or NewPrivateGroupChat.
+	notif, err = p.GetActivityCenterNotificationByID(nID3)
+	s.Require().NoError(err)
+	s.Require().False(notif.Deleted, notif.ID)
+	s.Require().False(notif.Dismissed, notif.ID)
+	s.Require().False(notif.Read, notif.ID)
```
Comment (seanstrom) [inline protocol/activity_center_persistence_test.go]: Here we're updating the original test case for deleting a message to not expect that we should notifications when they reference the deleted message via the `LastMessage` field.

File: protocol/activity_center_persistence_test.go [hunk 5]
```diff
@@ -205,17 +205,22 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
...
-		s.Require().True(notif.Read, notif.ID)
-	}
+	notif, err = p.GetActivityCenterNotificationByID(nID2)
+	s.Require().NoError(err, notif.ID)
+	s.Require().True(notif.Deleted, notif.ID)
+	s.Require().True(notif.Dismissed, notif.ID)
+	s.Require().True(notif.Read, notif.ID)
+
+	// Check that notifications with LastMessage.ID == messages[1].ID will remain.
+	// Unless the notification is of type NewOneToOne or NewPrivateGroupChat.
```
Comment (seanstrom) [inline protocol/activity_center_persistence_test.go]: TODO: We'll potentially need to update this comment

**Review by seanstrom — COMMENTED** [review]
Self Review 📖

File: protocol/activity_center_persistence_test.go [hunk 6]
```diff
@@ -205,17 +205,22 @@ func (s *ActivityCenterPersistenceTestSuite) Test_DeleteActivityCenterNotificati
...
 	s.Require().NoError(err)
 
-	for _, id := range []types.HexBytes{nID2, nID3} {
-		notif, err = p.GetActivityCenterNotificationByID(id)
-		s.Require().NoError(err, notif.ID)
-		s.Require().True(notif.Deleted, notif.ID)
-		s.Require().True(notif.Dismissed, notif.ID)
-		s.Require().True(notif.Read, notif.ID)
-	}
+	notif, err = p.GetActivityCenterNotificationByID(nID2)
```
Comment (igor-sirotin) [inline protocol/activity_center_persistence_test.go:213]: Also add this to prevent potential `panic` on access to `nil` pointer:
```go
s.Require.NotNil(notif)
```

File: protocol/activity_center_persistence.go (hunk 1 above)
Comment (igor-sirotin) [reply protocol/activity_center_persistence.go]: Looks to me like `LastMessage` can be completely removed 👍

File: protocol/activity_center_persistence_test.go (hunk 6 above)
Comment (igor-sirotin) [reply protocol/activity_center_persistence_test.go:213]: In fact, when `err != ni` , `notif` will be `nil`.
So this line is not quite good:
```go
s.Require().NoError(err, notif.ID)
```

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
Thanks for the tests! ❤️ 
I'm happy with the changes, just a minor thing in tests about accessing `nil`.

**Comment by seanstrom** [comment]
Awesome! thanks for the review @igor-sirotin 🙌 

I've updated the PR to remove the `LastMessage` checks, and I also updated the tests to guard against `nil` and avoid panics when inspecting the `err` value.

Let me know if these changes look good 🙏

**Review by ilmotta — APPROVED** [review]
LGTM!

Can we remove `LastMessage` entirely from the struct? The only places we set it now are when we unmarshal from the DB.

We would also need to create a migration to remove the column and update the mobile repo because we still have references to the last message. Does the desktop repo need adaptations? For these reasons it could be better to do the full clean up in a separate PR.

**Comment by seanstrom** [comment]
> Can we remove `LastMessage` entirely from the struct? The only places we set it now are when we unmarshal from the DB.
>

Yeah I'm not sure why notifications would need to reference the last message. It seems that `LastMessage` is intended to always be the last message of the related chat (the latest when retrieved?).

But when we access `:last-message` on mobile, it's rarely accessed directly from the notification. It usually seems to be accessed directly from the `chat` data.

For context, I think these are three main uses of `:last-message` with a notification:
* https://github.com/status-im/status-mobile/blob/97d4edcf30a1b34f48a6bfae80488cf496638bea/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L65-L68
* https://github.com/status-im/status-mobile/blob/97d4edcf30a1b34f48a6bfae80488cf496638bea/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L108-L111
* https://github.com/status-im/status-mobile/blob/97d4edcf30a1b34f48a6bfae80488cf496638bea/src/status_im/contexts/shell/activity_center/notification/contact_requests/view.cljs#L166-L170

Fortunately, each of those usages will default to using the `:message` field before defaulting the `:last-message` value. So we don't have to make any changes to mobile yet, but we would definitely need to clean those up soon.
 
> We would also need to create a migration to remove the column and update the mobile repo because we still have references to the last message. Does the desktop repo need adaptations? For these reasons it could be better to do the full clean up in a separate PR.

So there doesn't seem to be a column for `last_message` in the notifications table, because that data is joined from the `chat` table I think. So if did remove the struct field, we would have to double-check the places mobile or desktop directly access the field from a notification.

That all said, I'm all for a separate clean up PR 🧹 😄

**Comment by ilmotta** [comment]
> So there doesn't seem to be a column for `last_message` in the notifications table, because that data is joined from the `chat` table I think. So if did remove the struct field, we would have to double-check the places mobile or desktop directly access the field from a notification.

Indeed, the value is joined from the `chats` table 👍🏼

**Review by igor-sirotin — APPROVED** [review]
Thanks for the changes 👍

**Review by qfrank — APPROVED** [review]
Very detailed, thank you for the PR!

**Comment by churik** [comment]
@igor-sirotin @seanstrom 
are those breaking changes for the desktop app? If so, what is the effect?
We need to make status-desktop PR with required changes prior to merging, like we do this for mobile.

I'm not completely sure about procedure, in our case we just create status-mobile PR, based on particular status-go commit.

cc @anastasiyaig

**Comment by igor-sirotin** [comment]
@seanstrom @churik @anastasiyaig 

Yes, a status-desktop PR is required when doing a breaking change in status-go. 
This is the only way to make sure both clients are compatible with the change.
This is stated in the auto-comment above: https://github.com/status-im/status-go/pull/5789#issuecomment-2321433337

> are those breaking changes for the desktop app? If so, what is the effect?

I'm not sure what exactly is a breaking change here, but having `LastMessage` empty might break something indeed.

-----

I think the procedure should be:
1. PR in both clients MUST be opened, tested and approved before merging the status-go PR.
2. Communicate with desktop/mobile teams to make required changes 
(because desktop devs can't read clojure, and mobile devs can't read nim)

**Comment by seanstrom** [comment]
Yes I think desktop would be affected by these changes, but I haven't verified if desktop had the same glitch that was reported in https://github.com/status-im/status-mobile/issues/17052

The changes for removing `LastMessage` could create some issues, though it seems desktop checks for the `Message` field before checking for the `LastMessage` field here: https://github.com/status-im/status-desktop/blob/7c184f35002886947493070400500e7e05de74b6/src/app_service/service/activity_center/dto/notification.nim#L134-L138

---

I merged this PR too quickly, and in the future I'll ping in the status-go channel for both QA teams to see a potentially breaking change.
